### PR TITLE
add isTouching to Sprite and Group for evaluation purposes

### PR DESCRIPTION
**Pull request by cpirich**
@poorvasingal wants to test out a new `Sprite` and `Group` method called `isTouching`. It gives you the return value that you would receive if you called `bounce` or `collide` without actually changing any positions. Note that `overlap` can give you a similar return value, but it doesn't handle the tunneling cases properly.

Like the collision functions, `isTouching` resets the `touching` property for the calling sprite (not the target). The Group version resets the `touching` property for all of the sprites in the group. After calling `isTouching`, the return value will be either `true` or `false`. The `touching` property of sprites on the left-side (not the target(s)) will be updated so you can determine which sides touched: `left`, `right`, `top`, or `bottom`.

Unlike the collision functions, no positions will actually change. Also, there is no callback. Thus, it can be used like this:

```
if (sprite.isTouching(sprite2)) {
  sprite.bounce(sprite2);
}
```

However, the collision detection code will run twice in the case where `isTouching` returns `true`. We're going to put this in our version of gamelab for testing purposes. It is not in the palette or autocomplete at this time.

### CODE REVIEW

File: apps/src/gamelab/GameLabP5.js [hunk 1]
```diff
@@ -427,6 +428,22 @@ GameLabP5.prototype.init = function (options) {
...
+     * Test each member of group against the target using the isTouching sprite
+     * method.  Return true if any touching occurred.
+     *
+     * @method isTouching
+     * @param {Object} target Group or Sprite
+     * @return {boolean} True if any touching occurred
+     */
+    array.isTouching = function (target) {
+      var didTouch = false;
+      for (var i = 0; i < this.size(); i++) {
```
Comment (islemaster) [inline apps/src/gamelab/GameLabP5.js]: Any particular reason for `size()` instead of `length`?

File: apps/src/gamelab/GameLabP5.js [hunk 2]
```diff
@@ -427,6 +428,22 @@ GameLabP5.prototype.init = function (options) {
...
+        didTouch = this.get(i).isTouching(target) || didTouch;
```
Comment (islemaster) [inline apps/src/gamelab/GameLabP5.js]: Likewise any benefit to `get()` over array indexing `[]`?  I'm guessing it's because this is a 'copy' of collide() but since this is our own code I assume we can clean it up.

File: apps/src/gamelab/GameLabSprite.js [hunk 3]
```diff
@@ -300,3 +300,151 @@ module.exports.AABBops = function(p5Inst, type, target, callback) {
...
+  /*
+   * This code is a subset of the original AABBops method. It needs to be kept in
+   * sync with p5play's method, much like our stateful async version of AABBops
+   * above.
+   */
+
+  var abs = p5.prototype.abs;
+  var round = p5.prototype.round;
+  var quadTree = p5Inst.quadTree;
+  var createVector = p5Inst.createVector;
```
Comment (islemaster) [inline apps/src/gamelab/GameLabSprite.js]: Do these instance functions need to be bound?

File: apps/src/gamelab/GameLabP5.js (hunk 2 above)
Comment (islemaster) [inline apps/src/gamelab/GameLabP5.js]: Also, I was going to suggest an early-out here, but then I realized `isTouching()` has side effects on the sprites.  It might be worth a comment to that effect.

File: apps/test/npm-debug.log [hunk 4]
```diff
@@ -0,0 +1,28 @@
+0 info it worked if it ends with ok
```
Comment (islemaster) [inline apps/test/npm-debug.log]: You probably didn't mean to check in npm-debug.log, right?  We might need to put this in our .gitignore

File: apps/test/solutions/applab/ec_array.js [hunk 5]
```diff
@@ -0,0 +1,196 @@
+var testUtils = require('../../util/testUtils');
```
Comment (islemaster) [inline apps/test/solutions/applab/ec_array.js]: Is this test file related to the Gamelab change?  It doesn't seem to be.

File: apps/src/gamelab/GameLabP5.js (hunk 1 above)
Comment (cpirich) [inline apps/src/gamelab/GameLabP5.js]: This is just copying the _groupCollide() that you added to `p5.play.js` - I'm happy to change it!

File: apps/src/gamelab/GameLabP5.js (hunk 2 above)
Comment (cpirich) [inline apps/src/gamelab/GameLabP5.js]: I'll add a comment.

File: apps/test/npm-debug.log (hunk 4 above)
Comment (cpirich) [inline apps/test/npm-debug.log]: yeah, that was an accident.

File: apps/test/solutions/applab/ec_array.js (hunk 5 above)
Comment (cpirich) [inline apps/test/solutions/applab/ec_array.js]: and same with that.

File: apps/src/gamelab/GameLabSprite.js (hunk 3 above)
Comment (cpirich) [inline apps/src/gamelab/GameLabSprite.js]: Some aren't intentionally, copying what p5play does (but I added a comment for those).
Some should be: fixed in both places in this file (good catch!!)
Some aren't because they aren't called but are used for `instanceof` checks (added a comment)

File: apps/src/gamelab/GameLabP5.js [hunk 6]
```diff
@@ -427,6 +428,24 @@ GameLabP5.prototype.init = function (options) {
...
+     */
+    array.isTouching = function (target) {
+      var didTouch = false;
+      for (var i = 0; i < this.length; i++) {
+        didTouch = this[i].isTouching(target) || didTouch;
+        // Note: can't early out because each call needs to update the sprite's
+        // touching property.
+      }
+      return didTouch;
+    };
```
Comment (islemaster) [inline apps/src/gamelab/GameLabP5.js:447]: :+1:

**Comment by islemaster** [comment]
LGTM!
